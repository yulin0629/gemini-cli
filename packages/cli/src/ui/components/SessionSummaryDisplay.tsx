/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { Box, Text } from 'ink';
import Gradient from 'ink-gradient';
import { Colors } from '../colors.js';
import { formatDuration } from '../utils/formatters.js';
import { useSessionStats } from '../contexts/SessionContext.js';
import { computeSessionStats } from '../utils/computeStats.js';
import { FormattedStats, StatRow, StatsColumn } from './Stats.js';

// --- Prop and Data Structures ---

interface SessionSummaryDisplayProps {
  duration: string;
}

// --- Main Component ---

export const SessionSummaryDisplay: React.FC<SessionSummaryDisplayProps> = ({
  duration,
}) => {
  const { stats } = useSessionStats();
  const { metrics } = stats;
  const computed = computeSessionStats(metrics);

  const cumulativeFormatted: FormattedStats = {
    inputTokens: Object.values(metrics.models).reduce(
      (acc, model) => acc + model.tokens.prompt,
      0,
    ),
    outputTokens: Object.values(metrics.models).reduce(
      (acc, model) => acc + model.tokens.candidates,
      0,
    ),
    toolUseTokens: Object.values(metrics.models).reduce(
      (acc, model) => acc + model.tokens.tool,
      0,
    ),
    thoughtsTokens: Object.values(metrics.models).reduce(
      (acc, model) => acc + model.tokens.thoughts,
      0,
    ),
    cachedTokens: Object.values(metrics.models).reduce(
      (acc, model) => acc + model.tokens.cached,
      0,
    ),
    totalTokens: Object.values(metrics.models).reduce(
      (acc, model) => acc + model.tokens.total,
      0,
    ),
  };

  const totalRequests = Object.values(metrics.models).reduce(
    (acc, model) => acc + model.api.totalRequests,
    0,
  );

  const title = 'Agent powering down. Goodbye!';

  return (
    <Box
      borderStyle="round"
      borderColor="gray"
      flexDirection="column"
      paddingY={1}
      paddingX={2}
      alignSelf="flex-start"
    >
      <Box marginBottom={1} flexDirection="column">
        {Colors.GradientColors ? (
          <Gradient colors={Colors.GradientColors}>
            <Text bold>{title}</Text>
          </Gradient>
        ) : (
          <Text bold>{title}</Text>
        )}
      </Box>

      <Box marginTop={1}>
        <StatsColumn
          title={`Cumulative Stats (${totalRequests} API calls)`}
          stats={cumulativeFormatted}
          isCumulative={true}
        >
          <Box marginTop={1} flexDirection="column">
            <StatRow
              label="Total duration (API)"
              value={formatDuration(computed.totalApiTime)}
            />
            <StatRow
              label="Total duration (Tools)"
              value={formatDuration(computed.totalToolTime)}
            />
            <StatRow label="Total duration (wall)" value={duration} />
          </Box>
        </StatsColumn>
      </Box>
    </Box>
  );
};
