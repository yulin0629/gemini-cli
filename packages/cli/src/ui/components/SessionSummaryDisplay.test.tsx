/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import { render } from 'ink-testing-library';
import { describe, it, expect, vi } from 'vitest';
import { SessionSummaryDisplay } from './SessionSummaryDisplay.js';
import * as SessionContext from '../contexts/SessionContext.js';
import { SessionMetrics } from '../contexts/SessionContext.js';

vi.mock('../contexts/SessionContext.js', async (importOriginal) => {
  const actual = await importOriginal<typeof SessionContext>();
  return {
    ...actual,
    useSessionStats: vi.fn(),
  };
});

const useSessionStatsMock = vi.mocked(SessionContext.useSessionStats);

const renderWithMockedStats = (metrics: SessionMetrics) => {
  useSessionStatsMock.mockReturnValue({
    stats: {
      sessionStartTime: new Date(),
      metrics,
      lastPromptTokenCount: 0,
    },
  });

  return render(<SessionSummaryDisplay duration="1h 23m 45s" />);
};

describe('<SessionSummaryDisplay />', () => {
  it('correctly sums and displays stats from multiple models', () => {
    const metrics: SessionMetrics = {
      models: {
        'gemini-2.5-pro': {
          api: { totalRequests: 10, totalErrors: 1, totalLatencyMs: 50234 },
          tokens: {
            prompt: 1000,
            candidates: 2000,
            total: 3500,
            cached: 500,
            thoughts: 300,
            tool: 200,
          },
        },
        'gemini-2.5-flash': {
          api: { totalRequests: 5, totalErrors: 0, totalLatencyMs: 12345 },
          tokens: {
            prompt: 500,
            candidates: 1000,
            total: 1500,
            cached: 100,
            thoughts: 50,
            tool: 20,
          },
        },
      },
      tools: {
        totalCalls: 0,
        totalSuccess: 0,
        totalFail: 0,
        totalDurationMs: 0,
        totalDecisions: { accept: 0, reject: 0, modify: 0 },
        byName: {},
      },
    };

    const { lastFrame } = renderWithMockedStats(metrics);
    const output = lastFrame();

    // Verify totals are summed correctly
    expect(output).toContain('Cumulative Stats (15 API calls)');
    expect(output).toMatchSnapshot();
  });

  it('renders zero state correctly', () => {
    const zeroMetrics: SessionMetrics = {
      models: {},
      tools: {
        totalCalls: 0,
        totalSuccess: 0,
        totalFail: 0,
        totalDurationMs: 0,
        totalDecisions: { accept: 0, reject: 0, modify: 0 },
        byName: {},
      },
    };

    const { lastFrame } = renderWithMockedStats(zeroMetrics);
    expect(lastFrame()).toMatchSnapshot();
  });
});
