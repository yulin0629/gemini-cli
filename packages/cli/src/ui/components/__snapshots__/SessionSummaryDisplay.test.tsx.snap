// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SessionSummaryDisplay /> > correctly sums and displays stats from multiple models 1`] = `
"╭─────────────────────────────────────╮
│                                     │
│  Agent powering down. Goodbye!      │
│                                     │
│                                     │
│  Cumulative Stats (15 API calls)    │
│                                     │
│  Input Tokens                1,500  │
│  Output Tokens               3,000  │
│  Tool Use Tokens               220  │
│  Thoughts Tokens               350  │
│  Cached Tokens         600 (12.0%)  │
│  ─────────────────────────────────  │
│  Total Tokens                5,000  │
│                                     │
│  Total duration (API)        1m 2s  │
│  Total duration (Tools)         0s  │
│  Total duration (wall)  1h 23m 45s  │
│                                     │
╰─────────────────────────────────────╯"
`;

exports[`<SessionSummaryDisplay /> > renders zero state correctly 1`] = `
"╭─────────────────────────────────────╮
│                                     │
│  Agent powering down. Goodbye!      │
│                                     │
│                                     │
│  Cumulative Stats (0 API calls)     │
│                                     │
│  Input Tokens                    0  │
│  Output Tokens                   0  │
│  Thoughts Tokens                 0  │
│  ─────────────────────────────────  │
│  Total Tokens                    0  │
│                                     │
│  Total duration (API)           0s  │
│  Total duration (Tools)         0s  │
│  Total duration (wall)  1h 23m 45s  │
│                                     │
╰─────────────────────────────────────╯"
`;
